### PR TITLE
[kostalinverter] Add support for second generation inverters

### PR DESCRIPTION
**Pull request by basse04**
Örjan Backsell
orjan.backsell@gmail.com

This binding is redesigned to also work with the model New Generation of Kostal Solar Inverters.
There are possibilities to do an configuration for the unique channels that are interested to scrape from the Inverter. This option is only regarded to New Generation.
The original services are left as original, and there is possible to choose which model the binding will work for, due to config in demo.things.

Redesigned binding for Kostal Solar Inverters of model New Generation

[kostalinverter]

https://github.com/basse04/Kostal-Inverter-2.5.0/tree/master/org.openhab.binding.kostalinverter

https://github.com/basse04/Kostal-Inverter-2.5.0-SNAPSHOT

BR Basse

### CODE REVIEW

**Comment by Hilbrand** [comment]
Something went really wrong here. You have a copy of your binding in the root and also duplicated files in the binding itself. Also put all classes in the internal package, use the gson library part of the openHAB distribution and it's not very nice to remove the original authors name from the binding.xml. If you think this is a significant contribution to this binding append your name, and use your full name.

**Comment by cschneider** [comment]
I think your code might be based on an older version of the kostal inverter binding. Some of the newer changes seem to be lost after the PR. 
For example Christoph Weitkamp switched to the new unit types. There were also some channels that are now lost.
I think we should either find a way to make old and new hardware work with this smoothly or create a new binding for the new inverters.

**Comment by wborn** [comment]
There's also https://github.com/openhab/openhab2-addons/pull/4632 for adding support of several Kostal inverters. From an end user perspective it would make sense if all Kostal devices were supported with just one biding. WDYT @kaikreuzer ?

**Comment by RS84** [comment]
>           The problem is that I only have my device .. so it is difficult to understand the full spectrum of what Kostal offers.
> I think the common ground we have is that we scrape the web UI. So I think it would help if everyone who owns a kostal inverter posts the html the UI provides .. then we can find the commonalities.

There are 3 generations of KOSTAL Inverters currently (PIKO MP plus excluded; this was not originally developed by KOSTAL)
For our purpose (consuming the data) they differ in the used communication board.
Generation 1 & 2 are based on an own operating System offering limited access to data (except for parsing the html files provided; which the kostalinverter-addon does (not sure for which generation))
The 3. generation is based on a linux based communication board. This board has way more power than the older once. The web page is modern and based on JavaScript, css, aaaand in the backend: a REST api :-)
This API is not officially anounced yet, but having a talk with the developers they told me that this API is stable. The documentation is missing at the moment.

The mentioned PR https://github.com/openhab/openhab2-addons/pull/4632 (I´m the author btw) is based on this REST api

**Comment by Hilbrand** [comment]
@basse04 Thanks for the contribution and also fixing the binding author. I was wondering what your plans are with working on the binding. I'm asking this because there are a number of issues with your implementation that might require significant changes before this binding will be merged. Some of the more structural  issues:
* This binding completely changes the channels with regards to the original binding. This means it would break all existing usages. This should not be done unless there are good reasons for it.
* You have modeled your channels to be very generic and the binding requires a special file to configure. While this is kind of flexible, it makes it impossible to make use of the QuantityTypes. The special configuration file also means the user must do some additional configuration that is different than the rest of openHAB. This should be avoided. The channels should be designed from the user point of view. This means all possible type channels should be added. If the channels would need a configurable id (which is specified in your config file) it should be a property of the channel not in a dedicated file.
* There are a lot of files added in the root that should not be there.

So to summarize this binding needs a lot of (re)work to get ready to be integrated, almost to the point I would just recommend to start over and re-add your changes to the current binding. It that something you want to do? Because if so I think we should agree what to do with this pr? We could close it for now if it's to much trouble to get it right. I can also give you some tips/ideas on how to redesign and setup the binding for both old and new models (we can start a topic on the forum about this). You might contact @RS84 to see if you can work together as the other pr also covers these products and add also more channels. Let us know what you think.

**Comment by Hilbrand** [comment]
@basse04 Is this pr still relevant with all new models added in the already merged pr #4632?  Or can this pr be closed?

**Comment by martinvw** [comment]
> @basse04 Is this pr still relevant with all new models added in the already merged pr #4632? Or can this pr be closed?

Hello @basse04 what is the status of this pull request, could you please give us an update?

**Comment by Hilbrand** [comment]
This or is replaced by #6450 (and also has no remote branch anymore)
